components/gamification/store/StoreItemsList.tsx

```diff
@@ -0,0 +1,137 @@
+import { useCallback, useState } from 'react';
+import { FlatList, KeyboardAvoidingView, Platform, RefreshControl, StyleSheet, Text, View } from 'react-native';
+import CenteredFeedback from '../../common/feedback/CenteredFeedback';
+import StoreItemCard from './StoreItemCard';
+import { useGetStoreItemsQuery, useGetTransactionQuery, useGetUserItemsQuery, usePurchaseItemMutation } from '../api';
+import { IStoreItem, IStoreItemWithQuantity } from '../types';
+import FetchError from '../../common/feedback/FetchError';
+import LoadingIndicator from '../../common/feedback/LoadingIndicator';
+import ShimmerPlaceholder from 'react-native-shimmer-placeholder';
+import { LinearGradient } from 'expo-linear-gradient';
+import { generateErrorResponseMessage } from '../../../utils/httpUtils';
+import useNotifications from '../../../hooks/useNotifications';
+import GemsIndicatorButton from '../../transaction/GemsIndicatorButton';
+import { isDataResponse } from '../../../services';
+
+
+const StoreItemsList = () => {
+  const { data: storeItemsPage, isLoading: isStoreItemsLoading, isError: isStoreItemsError, refetch: storeItemsRefetch } = useGetStoreItemsQuery();
+  const { data: userItemsPage, isLoading: isUserItemsLoading, isError: isUserItemsError, refetch: userItemsRefetch } = useGetUserItemsQuery();
+  const { data: userGemsData, isLoading: isUserGemsLoading, isError: isUserGemsError, refetch: userGemsRefetch } = useGetTransactionQuery();
+  const [purchase, { isError: isPurchaseError, error: purchaseError }] = usePurchaseItemMutation();
+  const [isRefreshing, setIsRefreshing] = useState(false);
+  const { add } = useNotifications();
+
+  const onRefresh = useCallback(async () => {
+    setIsRefreshing(true);
+    await storeItemsRefetch();
+    await userItemsRefetch();
+    await userGemsRefetch();
+    setIsRefreshing(false);
+  }, [storeItemsRefetch, userItemsRefetch, userGemsRefetch]);
+
+  const renderSkeletonList = () => {
+    return (
+      <View style={styles.skeletonContainer}>
+        {Array.from({ length: 6 }).map((_, index) => (
+          <ShimmerPlaceholder key={index} style={styles.skeletonRectangle} LinearGradient={LinearGradient} />
+        ))}
+      </View>
+    );
+  };
+
+  if (isStoreItemsLoading || isUserItemsLoading || isUserGemsLoading) return renderSkeletonList();
+
+  if (isStoreItemsError || isUserItemsError || isUserGemsError || !storeItemsPage || !userItemsPage || !userGemsData) {
+    return <FetchError />;
+  }
+
+  if (!storeItemsPage?.storeItems || storeItemsPage?.storeItems?.length === 0) {
+    return (
+      <CenteredFeedback message="There are no items in the store" />
+    );
+  }
+
+  const handleGemsPress = async (item: IStoreItemWithQuantity) => {
+    const purchaseResponse  = await purchase({ itemId: item.id});
+    if (purchaseError || isPurchaseError) {
+      add({
+        body: generateErrorResponseMessage(purchaseError),
+        type: 'error',
+      });
+      return;
+    }
```

Comment:
There is an issue here I could not figure out. If the user does not have enough gems, then the purchase endpoint returns an error with a message. However, when the user first taps the button, the error message is not shown on the screen (even though condole.log(purchaseResponse) shows that there is an error). 

If the user taps again, only then the correct message is shown (and for all subsequent attempts as well).

Reply:
This will not work at all, you need to use the `useError` hook whenever you need to generate an error message received from the API @KardelenCeren

Reply:
Thanks for the fix! I also added a check for isDataResponse here - otherwise, if the user tapped many times on the button a successful message was shown even though the user did not have any gems.